src/dma.rs

```diff
@@ -202,25 +182,87 @@ impl XdmacChannel {
     }
 
     pub fn set_bi_interrupt(&self, enable: bool) {
-        let mut dma = CSR::new(self.xdmac_base_addr as *mut u32);
-
-        let (cie_reg, cid_reg) = match self.channel {
-            DmaChannel::Channel0 => (XDMAC_CIE0, XDMAC_CID0),
-            DmaChannel::Channel1 => (XDMAC_CIE1, XDMAC_CID1),
-        };
+        const MASK_BI: u32 = 1 << 0;
 
         if enable {
-            dma.wo(cie_reg, 1 << 0); // BI
+            self.set_cie(MASK_BI);
         } else {
-            dma.wo(cid_reg, 1 << 0); // BI
+            self.set_cid(MASK_BI);
         }
     }
+
+    /// Sets the value of the `CC` register for this channel.
+    fn set_cc(&self, cc_val: u32) {
+        const CC_REG_OFFSET: u32 = 0x78;
+        let cc_reg_ptr = self.reg_by_offset(CC_REG_OFFSET);
+        unsafe {
+            cc_reg_ptr.write_volatile(cc_val);
+        }
+    }
+
+    /// Sets the value of the `CIE` (channel interrupt enable) register.
+    fn set_cie(&self, cie_val: u32) {
+        const CIE_REG_OFFSET: u32 = 0x50;
+        let cc_reg_ptr = self.reg_by_offset(CIE_REG_OFFSET);
+        unsafe {
+            cc_reg_ptr.write_volatile(cie_val);
+        }
+    }
+
+    /// Sets the value of the `CID` (channel interrupt disable) register.
+    fn set_cid(&self, cid_val: u32) {
+        const CID_REG_OFFSET: u32 = 0x54;
+        let cc_reg_ptr = self.reg_by_offset(CID_REG_OFFSET);
+        unsafe {
+            cc_reg_ptr.write_volatile(cid_val);
+        }
+    }
+
+    /// Sets channel's source and destination addresses.
+    fn set_sa_da(&self, sa: u32, da: u32) {
+        const CSA_REG_OFFSET: u32 = 0x60;
+        const CDA_REG_OFFSET: u32 = 0x64;
+        let sa_reg_ptr = self.reg_by_offset(CSA_REG_OFFSET);
+        let da_reg_ptr = self.reg_by_offset(CDA_REG_OFFSET);
+        unsafe {
+            sa_reg_ptr.write_volatile(sa);
+            da_reg_ptr.write_volatile(da);
+        }
+    }
+
+    fn set_data_size(&self, size: u32) {
+        const CUBC_REG_OFFSET: u32 = 0x70;
+        let ubl_reg_ptr = self.reg_by_offset(CUBC_REG_OFFSET);
+        unsafe {
+            ubl_reg_ptr.write_volatile(size);
+        }
+    }
+
+    fn reg_by_offset(&self, offset: u32) -> *mut u32 {
+        let reg_addr =
+            self.xdmac_base_addr + offset + self.channel as u32 * DMA_CHANNEL_NUM_REGISTERS;
+        reg_addr as *mut u32
+    }
 }
 
 #[derive(Debug, Copy, Clone)]
 pub enum DmaChannel {
     Channel0 = 0,
```

Comment:
Note: at this point I'm not sure I'd bother with an enum. Especially since in the future we might want to assign dma channels dynamically at runtime.